Feat project table schema

Benjtalkshow opened the pull request:
Create Projects Table Schema

Closes #138
This PR implements the core `projects` table schema for kindfi according to the requirements. The table is designed to store essential project information while maintaining data integrity through constraints and proper security through RLS policies.

## Changes Implemented

1. Created the `projects` table with all required fields:
   - Basic project information (id, title, description)
   - Financial data (current_amount, target_amount, min_investment)
   - Status tracking (percentage_complete, investors_count)
   - Metadata (created_at, updated_at)
   - Category reference (as TEXT type for now)
   - Image URL field
   - Owner reference
   
2. Added appropriate constraints:
   - Primary key on id
   - Check constraints for business rules
   - Foreign key to auth.users for owner_id
   
3. Implemented RLS policies:
   - Project creation limited to authenticated users
   - Updates limited to project owners
   - Read access available to everyone
   
4. Created necessary indices for performance optimization
5. Set up trigger for automatic timestamp updates

## Notes

- The `category_id` field is implemented as TEXT type as recommended, pending the future creation of the categories table
- Relationships with other tables (investors, tags, updates)




## Screenshots
![image](https://github.com/user-attachments/assets/6f7505c6-b687-4886-9648-0b857c5b60e5)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced investment project management with extended project details and improved tracking of funding progress.
  - Introduced secure escrow and contributions functionality to support streamlined project funding.

- **Documentation**
  - Added comprehensive documentation outlining project data structures, security policies, and usage guidelines for managing investments.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Benjtalkshow commented:
@AndlerRL 
I was wondering why i have 15k+ changes. It seems when i run `bun install` it triggered the execution of some contract files or so. Just seeing it now.

Benjtalkshow commented on services/supabase/package.json:
```diff
@@ -22,9 +22,7 @@
 	"devDependencies": {
-		"supabase-to-zod": "^1.0.7",
-		"supabase": "^2.0.0"
+		"supabase-to-zod": "^1.0.7"
```
Alright will fix this

AndlerRL replied:
@coderabbitai can you check if we have the dependencies missing? This errors looks we are :grimacing: first time seeing this in the vercel builds

AndlerRL replied:
@coderabbitai I meant with the error in the screenshot, about the husky dependency but that is ok, you highlighted a good point too that I was mentioning too.

AndlerRL commented on package-lock.json:
It is safe to remove this file, please remove.

AndlerRL reviewed (changes requested):
Good changes! I see there is only a few clean up before to do the merge :smiley:

After handling those, it will be good to go!

AndlerRL commented:
@Benjtalkshow I see these conflicts now after merging PR #161 . This is how to fix it:

![image](https://github.com/user-attachments/assets/d0d27a09-7407-4fbc-bbe7-a4ea4a14c8d9)


1. Run `bun install` either at root or `supabase` service folder.
2. Do `bun db:push` to push your changes with the new changes.
3. Run the script `bun gen` to fix the conflicts.

Benjtalkshow commented:
> @Benjtalkshow I see these conflicts now after merging PR #161 . This is how to fix it:
> 
> ![image](https://private-user-images.githubusercontent.com/35474730/414170410-d0d27a09-7407-4fbc-bbe7-a4ea4a14c8d9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk4NjkyMzYsIm5iZiI6MTczOTg2ODkzNiwicGF0aCI6Ii8zNTQ3NDczMC80MTQxNzA0MTAtZDBkMjdhMDktNzQwNy00ZmJjLWJiZTctYTRlYTRhMTRjOGQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE4VDA4NTUzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkZTE0Mzk2YWM0YjIzNjExNDcwOGIzNDc1MmQyMjY1ZDAxZjk1NTIxZThkZTg0YmVlY2M4YmJkOWI2NjRkNzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Yry0Ss7gp3est982cCgNEa76lprWUXh8KN_iO0cysX4)
> 
> 1. Run `bun install` either at root or `supabase` service folder.
> 2. Do `bun db:push` to push your changes with the new changes.
> 3. Run the script `bun gen` to fix the conflicts.

@AndlerRL What about the supabase bot comment saying "no changes deteced"? should i create a new sub "supabase" folder and move my migration file inside?

AndlerRL commented:
It's ok, @Benjtalkshow . As soon as I saw the comment I updated the supabase ci/cd config. It should read on new commits the updated root folder

AndlerRL commented:
@Benjtalkshow I think you will have to update a little your migrations. There are some differences between the PR #161 which are minimal in comparison. You will probably delete the name field at the end. I am sharing the @Jagadeeshftw table version on the related supabase preview:

![image](https://github.com/user-attachments/assets/8a656fb9-9988-4bf5-8c04-d1d52b5cf7f7)

Benjtalkshow commented:
> @Benjtalkshow I think you will have to update a little your migrations. There are some differences between the PR #161 which are minimal in comparison. You will probably delete the name field at the end. I am sharing the @Jagadeeshftw table version on the related supabase preview:
> 
> ![image](https://private-user-images.githubusercontent.com/35474730/414672646-8a656fb9-9988-4bf5-8c04-d1d52b5cf7f7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk5OTMxMzksIm5iZiI6MTczOTk5MjgzOSwicGF0aCI6Ii8zNTQ3NDczMC80MTQ2NzI2NDYtOGE2NTZmYjktOTk4OC00YmY1LThjMDQtZDFkNTJiNWNmN2Y3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE5VDE5MjAzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFhZGJlNDA4YjIyMGRhYWFmMTZhYTg3NGUzMGEyMDYzNDYyNmFiYzMyYTkzYTg1NmQ5NzQxZDEyNmQzNzJhOWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.m4l7dhU1JjNOasSveg0J5OEMNhh_G15dvaxETd7knXU)
> 
> PD: If you have an auto-generated file conflict, re-generate them while having the local db running. That should fix the conflict issue.

ok

Benjtalkshow commented:
> @Benjtalkshow I think you will have to update a little your migrations. There are some differences between the PR #161 which are minimal in comparison. You will probably delete the name field at the end. I am sharing the @Jagadeeshftw table version on the related supabase preview:
> 
> ![image](https://private-user-images.githubusercontent.com/35474730/414672646-8a656fb9-9988-4bf5-8c04-d1d52b5cf7f7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk5OTM0NjksIm5iZiI6MTczOTk5MzE2OSwicGF0aCI6Ii8zNTQ3NDczMC80MTQ2NzI2NDYtOGE2NTZmYjktOTk4OC00YmY1LThjMDQtZDFkNTJiNWNmN2Y3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE5VDE5MjYwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM4NjAxMWI2NjhiYmY2NjFkYjQzMTZiMTk1NWJmZjViNmRlNTM1N2VmNThiYTJiMTIxM2I1MzkxYTY3ZWVmNzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.y5UL2Eeo3HiEqyTb0hmwW3GmtqE14nueKhpG2SZcSNc)
> 
> PD: If you have an auto-generated file conflict, re-generate them while having the local db running. That should fix the conflict issue.

@AndlerRL I dont really understand this correction. can you explain more?

Benjtalkshow commented:
@AndlerRL 
I'm working with the requirements outlined in my issue description. Are you suggesting I modify my script to align with @Jagadeeshftw , or should I remove his project table schema entirely and link it to my already created project schema instead?

AndlerRL commented:
Sure no problem @Benjtalkshow. @Jagadeeshftw had the task to develop the escrow off-chain tracking, which needed to have a relationship with the projects table hence, he did a basic project table in order to complete his task.

Since this PR deals with the project and, your migration is after Jagadeesh implementation, it might have migration issues and in order to know that it will be applying your migration with the changes in develop. If it doesn't happen (the migration error) but you see extra fields, you can remove them. This probably would cause another migration but we will be sure about it after applying it.

Let me know if you need further assistance, I can take this so you can focus only on the task while I can assist in finalizing the connection (which is what we need to do now). My ETA is in 5hrs if you haven't checked this yet.

Benjtalkshow commented:
@AndlerRL 
The changes i made ensures that the `project_id references` in the `create_escrow_contracts` schema correctly use the UUID from the main projects table.

### Changes Made:
1. Deleted the projects table in the `create_escrow_contracts` schema to avoid duplication.
2. Modified `project_id` references to `TEXT` in `create_escrow_contracts` to allow for a smooth migration.
3. Pushed the updated database schema to ensure consistency.
4. Altered all `project_id` references back to `UUID` after ensuring the main projects table was created correctly.
5. Lastly updated the `create_escrow_contracts` schema to use projects_id referencing the main project table.

AndlerRL commented:
> @AndlerRL The changes i made ensures that the `project_id references` in the `create_escrow_contracts` schema correctly use the UUID from the main projects table.
> 
> ### Changes Made:
> 1. Deleted the projects table in the `create_escrow_contracts` schema to avoid duplication.
> 2. Modified `project_id` references to `TEXT` in `create_escrow_contracts` to allow for a smooth migration.
> 3. Pushed the updated database schema to ensure consistency.
> 4. Altered all `project_id` references back to `UUID` after ensuring the main projects table was created correctly.
> 5. Lastly updated the `create_escrow_contracts` schema to use projects_id referencing the main project table.

Excellent work, @Benjtalkshow  :+1: yes, that was it! I made some tweaks before the push to develop. Thank you, the DB schema is taking body now! :muscle:

AndlerRL reviewed (approved):
LGTM! Thanks for dealing with the details, good documentation as well :ok_hand: I made few tweaks for DevOps and in the migrations orders. I was making sure to have the migrations in order i.e.: if a table refers another, that referred table should be created before and the property exist.

Merging now!

Benjtalkshow commented:
> LGTM! Thanks for dealing with the details, good documentation as well 👌 I made few tweaks for DevOps and in the migrations orders. I was making sure to have the migrations in order i.e.: if a table refers another, that referred table should be created before and the property exist.
> 
> Merging now!

Thats good. Thank you @AndlerRL
